File: GraphAlgorithms/DFS.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

// Function to perform DFS traversal
void dfsTraversal(vector<vector<int>>& graph, int start, vector<bool>& visited) {
    stack<int> stk;
    stk.push(start);

    while (!stk.empty()) {
        int curr = stk.top();
        stk.pop();

        if (visited[curr]) {
            continue;
        }

        visited[curr] = true;
        cout << curr << " "; // Process the current vertex

        for (int neighbor : graph[curr]) {
            if (!visited[neighbor]) {
                stk.push(neighbor);
            }
        }
    }
}

// Function to initialize and call DFS traversal
void dfs(vector<vector<int>>& graph, int start) {
    int n = graph.size(); // Number of vertices in the graph
    vector<bool> visited(n, false); // Array to track visited vertices

    cout << "DFS traversal: ";
    dfsTraversal(graph, start, visited);
}

int main() {
    int n; // Number of vertices
    int m; // Number of edges

    n = 6; // Number of vertices in the graph
    m = 7; // Number of edges in the graph

    vector<vector<int>> graph(n); // Adjacency list

    // Define the edges of the graph
    graph[0] = {1, 2};
    graph[1] = {0, 3, 4};
    graph[2] = {0, 4};
    graph[3] = {1, 4, 5};
    graph[4] = {1, 2, 3, 5};
    graph[5] = {3, 4};

    int startVertex = 0; // Starting vertex for DFS

    dfs(graph, startVertex);

    return 0;
}

File: GraphAlgorithms/DFS.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

// Graph class
class Graph {
    int numNodes;
    private List<List<Integer>> adjList;

    public Graph(int numNodes) {
        this.numNodes = numNodes;
        adjList = new ArrayList<>();
        for (int i = 0; i < numNodes; i++) {
            adjList.add(new ArrayList<>());
        }
    }

    public void addEdge(int src, int dest) {
        adjList.get(src).add(dest);
    }

    public List<Integer> getAdjacentNodes(int node) {
        return adjList.get(node);
    }
}

// DFS algorithm
class DFS {
    public static void dfs(Graph graph, int startNode, int target) {
        int numNodes = graph.numNodes;
        boolean[] visited = new boolean[numNodes];
        Stack<Integer> stack = new Stack<>();

        // Mark the start node as visited and push it to the stack
        visited[startNode] = true;
        stack.push(startNode);

        while (!stack.isEmpty()) {
            int node = stack.pop();
            System.out.print(node + " ");  // Process or visit the current node

            if (node == target) {
                System.out.println("\nTarget found!");
                return;
            }

            // Get the adjacent nodes of the current node
            List<Integer> neighbors = graph.getAdjacentNodes(node);

            // Visit unvisited neighbors and push them to the stack
            for (int neighbor : neighbors) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    stack.push(neighbor);
                }
            }
        }

        System.out.println("\nTarget not found!");
    }
// }

// // Main class
// public class Main {
    public static void main(String[] args) {
        int numNodes = 8; // Number of nodes in the graph

        Graph graph = new Graph(numNodes);

        // Add edges to the graph
        graph.addEdge(0, 1);
        graph.addEdge(0, 2);
        graph.addEdge(1, 3);
        graph.addEdge(1, 4);
        graph.addEdge(2, 5);
        graph.addEdge(2, 6);
        graph.addEdge(3, 7);

        int startNode = 0; // Starting node for DFS
        int target = 7; // Target number to search

        System.out.print("DFS Traversal: ");
        DFS.dfs(graph, startNode, target);
    }
}
